src/wasm.h

```diff
@@ -1172,7 +1172,39 @@ struct BinaryLocations {
   struct Span {
     BinaryLocation start, end;
   };
+
+  // Track the range of addresses an expressions appears at. This is the
+  // contiguous range that all instructions have - control flow instructions
+  // have additional opcodes later (like an end for a block or loop), see
+  // just after this.
   std::unordered_map<Expression*, Span> expressions;
+
+  // Track the extra offsets that some instructions, in particular control
+  // flow, have, like 'end' for loop and block. We keep these in a separate map
+  // because they are rare and we optimize for the storage space for the common
+  // type of instruction which just needs a Span.
+  // We implement this as a simple struct with two elements (as two extra
+  // elements is the maximum currently needed; due to 'catch' and 'end' for
+  // try-catch). The second value may be 0, indicating it is not used.
+  struct ExtraLocations : public std::array<BinaryLocation, 2> {
+    ExtraLocations() {
+      // Ensure zero-initialization.
+      for (auto& item : *this) {
+        item = 0;
+      }
+    }
+  };
+
+  enum ExtraId {
+    // All control flow structures have an end, so use index 0 for that.
+    End = 0,
+    // Use index 1 for all other current things.
+    Else = 1,
+    Catch = 1,
+    Invalid = -1
+  };
+  std::unordered_map<Expression*, ExtraLocations> extraExpressions;
```

Comment:
this one is especially confusing because while there are extra instructions in the binary, they are not extra *expressions* in the Binaryen sense.

edit: actually this is a mapping from expressions, not to expressions. what's the difference between this and `extraExpressionLocations` below?

Reply:
Correct, this is meant to be "extra" for the expressions item earlier. Perhaps it should be expressionsExtra? or extraForExpressions? (naming is hard...)

The difference between this and `extraExpressionLocations` is that one is for reading and one for writing. For reading, we store the information on the Function object (where it can then be processed in parallel, after reading). For writing, we store it on the Module object. I tried to keep a consistent naming:

 * On a BinaryLocations object it is clear we are talking about a location, so just `.expressions` means "locations for expressions".
 * On a Function object, among other function-related fields, the name is longer, `expressionLocations`, to indicate it is about locations.

(Another option might be to move the function fields into a containing structure; I don't feel strongly either way.)